plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftUnloadSplitManager.java

```diff
@@ -0,0 +1,200 @@
+                        jdbcTableHandle.getRequiredNamedRelation().getRemoteTableName()));
+
+        ConnectorSplitSource fallbackSplitSource = new FixedSplitSource(new JdbcSplit(Optional.empty()));
+
+        // Fallback to jdbc for no table columns in projection
+        if (((JdbcTableHandle) table).getColumns().isPresent() && ((JdbcTableHandle) table).getColumns().get().isEmpty()) {
+            return fallbackSplitSource;
+        }
+        if (containsUnsupportedType(columns)) {
+            return fallbackSplitSource;
+        }
+        if (((JdbcTableHandle) table).getLimit().isPresent()) {
+            return fallbackSplitSource;
+        }
+        if (containsFilterConditionWithDecimalTypeColumn(table)) {
+            return fallbackSplitSource;
+        }
+        Connection connection;
+        PreparedStatement statement;
+        try {
+            connection = jdbcClient.getConnection(session);
+            statement = buildUnloadSql(session, connection, jdbcTableHandle, columns);
+        }
+        catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+        return new RedshiftUnloadSplitSource(executor, connection, statement, fallbackSplitSource);
+    }
+
+    private PreparedStatement buildUnloadSql(ConnectorSession session, Connection connection, JdbcTableHandle table, List<JdbcColumnHandle> columns)
+            throws SQLException
+    {
+        PreparedQuery preparedQuery = jdbcClient.prepareQuery(session, table, Optional.empty(), columns, ImmutableMap.of());
+
+        PreparedStatement openTelemetryPreparedStatement = queryBuilder.prepareStatement(jdbcClient, session, connection, preparedQuery, Optional.of(columns.size()));
+        RedshiftPreparedStatement redshiftPreparedStatement;
+        CachedQuery redshiftPreparedQuery;
+        ParameterList preparedParameters;
+        try {
+            Field preparedStatementField = OpenTelemetryStatement.class.getDeclaredField("delegate");
+            preparedStatementField.setAccessible(true);
+            redshiftPreparedStatement = (RedshiftPreparedStatement) preparedStatementField.get(openTelemetryPreparedStatement);
+
+            Field preparedQueryField = RedshiftPreparedStatement.class.getDeclaredField("preparedQuery");
+            preparedQueryField.setAccessible(true);
+            redshiftPreparedQuery = (CachedQuery) preparedQueryField.get(redshiftPreparedStatement);
+
+            Field preparedParametersField = RedshiftPreparedStatement.class.getDeclaredField("preparedParameters");
+            preparedParametersField.setAccessible(true);
+            preparedParameters = (ParameterList) preparedParametersField.get(redshiftPreparedStatement);
+        }
+        catch (IllegalAccessException | NoSuchFieldException e) {
+            throw new RuntimeException(e);
+        }
+
+        openTelemetryPreparedStatement.close();
+
+        String selectQuerySql = redshiftPreparedQuery.query.toString(preparedParameters);
```

Comment:
We would be loosing the SQLInjection here right ?

Reply:
I don't think so as it uses PreparedStatement to get Redshift query with parameters placed in which is then used in unload sql